Fixed error in SUR objective and numerical issues.

mirhahn opened the pull request:
Changes:
    - Reported rounding objective in SUR is now maximum over
      cumulative error at all times rather than only at end of
      time horizon.
    - Removed "sum equals one" check in BinApprox due to numerical
      issues with floating point numbers returned by relaxation solver.
    - Replaced default values for max uptimes with infinity to avoid
      triggering constraints due to numerical errors.

mirhahn committed:
Fixed error in SUR objective and numerical issues.

Changes:
    - Reported rounding objective in SUR is now maximum over
      cumulative error at all times rather than only at end of
      time horizon.
    - Removed "sum equals one" check in BinApprox due to numerical
      issues with floating point numbers returned by relaxation solver.
    - Replaced default values for max uptimes with infinity to avoid
      triggering constraints due to numerical errors.

mirhahn commented:
Fails integration tests due to removed SOS1 check. SOS1 check should not be implemented using floating point equality check. Even adjusting the values (e.g. by setting one of the controls to "1.0 - sum of other controls") may not satisfy this condition due to rounding errors.

adbuerger commented:
Hi @mirhahn , thank you for your pull request!

Regarding the "sum equals one" check, I think we cannot just remove it since pycombina, even if we removed the functionality that pycombina automatically adds an "off state", we should somehow check whether the input provided by the user makes sense.

I suggest we introduce some tolerance for this check, and add the possibility to turn the check off in case this should be requested by a user. However, by default the check should be active.

What do you think?

mirhahn commented:
Hi @adbuerger,

There might be ways to make the SOS1 check account for certain inaccuracies. The problem is always that there are sources of error unknown to Combina. Even if the sources of error are known, their magnitude can be hardware dependent. We could make some worst-case estimates using machine epsilon. We can get that from `numpy.finfo(float).eps`.

The `sum == 1.0` check is definitely a bad idea. Even if I pick `n - 1` numbers between `0.0` and `1.0` such that they sum up to less than `1.0` and set the last number to `1.0 - sum(numbers[:n-1])`, it can still fail because sums and differences are not accurate to that degree. If we apply a fixed tolerance, there has to be a way to circumvent the check when working with iterative optimization code which has relatively lax primal feasibility enforcement.

It might be possible to add an optional argument (with some arbitrary default value like `1e-9`) to the constructor that lets the user specify the desired accuracy. Then again, the only type of user who would ever change the value of this bound to something more meaningful is someone who would also make sure that their relaxed controls add up to `1.0` (within reason). That might make the whole thing pointless.

adbuerger commented:
Thank you for your explanation @mirhahn, I think the following might make sense, then:

* We further assume that the "off state" is always included in the data passed to pycombina, and remove the functionality of automatically inserting such an additional switch.
* For the relaxed binary data passed, we check their sum at each time point. If the values are below, say, 0.995, we raise a warning (not an error) that states this circumstance, and that the user should check the inputs to assure that they're correct.

Am I missing something here?

mirhahn committed:
Merge remote-tracking branch 'upstream/develop' into develop

mirhahn committed:
Improved SOS1 check.

Changes:
    - SOS1 check now has very basic dynamic tolerance to account for
      numerical errors incurred due to floating point arithmetic and
      binary threshold handling.
    - SOS1 check now raises a warning instead of an exception.
    - Modified input unit tests to check for warnings rather than
      exceptions.
    - Added two new randomized unit tests to stress test numerical
      tolerances for two normalization strategies.

mirhahn commented:
Hi @adbuerger,

I think this solution is a good first approach. There's now a dynamic tolerance (that is rather generous) in the SOS1 check that should account for errors that occur both due to inaccurate addition and due to errors accumulated when controls close to 0.0 or 1.0 are clamped to those respective values. The code now issues a warning rather than raising an exception. I have also taken the liberty of adjusting the corresponding unit test accordingly and adding two new unit tests of my own that emulate normalization strategies that the user might use. They use randomized data, so they're not 100 per cent accurate. It might be difficult to come up with fixed data that causes numerical issues on all machines.

The code is already merged with the fix for #5 so it should be up to date with your development branch.

mirhahn committed:
Tightened error margins on SOS1 check.

Changes:
    - When clamping controls to 0.0 or 1.0, BinaryApproximation now
      keeps track of the number of clamped controls per time interval.
      This data is then used to adjust SOS1 tolerances using a tighter
      error margin. The resulting SOS1 check strikes a good balance
      between performance-hungry methods that calculate the actual
      difference made by the adjustment and fast methods that assume
      that every control value might be clamped.

adbuerger commented:
Great, thank you @mirhahn for your contribution!
